Do not mask node operation errors

bhalevy opened the pull request:
This series handles errors when aborting node operations and prints them rather letting them leak and be exposed to the user.

Also, cleanup the node_ops logging formats when aborting different node ops
and add more error logging around errors in the "worker" nodes.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4286/

denesb commented on service/storage_service.cc:
```diff
@@ -2079,7 +2080,7 @@ future<> storage_service::decommission() {
                 slogger.info("decommission[{}]: Finished token ring movement, removing node={}, sync_nodes={}, ignore_nodes={}", uuid, endpoint, nodes, ignore_nodes);
             } catch (...) {
-                slogger.warn("decommission[{}]: Abort decommission operation started, removing node={}, sync_nodes={}, ignore_nodes={}", uuid, endpoint, nodes, ignore_nodes);
+                slogger.error("decommission[{}]: decommissioning node={}, sync_nodes={}, ignore_nodes={} failed: {}. Aborting operation", uuid, endpoint, nodes, ignore_nodes, std::current_exception());
```
In fact this entire catch section looks like it is very similar across the different ops, maybe extract it into a function?

bhalevy replied:
Hmm, maybe this could work of each node op prints its own message, since they have slightly different parameters, but it's not a bad idea.

bhalevy commented:
In v2 ([35ea9da](https://github.com/scylladb/scylladb/commit/35ea9dab31a8fc661bab682894ab7e82e3fcc826)):
- storage_service: add node_ops_ctl class to formalize all node_ops flow
  - overhaul of node_ops coordinator control path using common code for sending each cmd and handling errors.

bhalevy commented:
In v3 ([655300a](https://github.com/scylladb/scylladb/commit/655300adb975880432ec52038eaebbb15f1b5957)):
- storage_service: add node_ops_ctl class to formalize all node_ops flow
  - added abort_on_error that just prints errors from the abort cmd and ignores them, not to mask the main error causing the abort
  - plain abort() method left in anticipation to integrating node_ops with task_manager by @Deexie and aborting cleanly when the node operation is aborted, once we support it.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4297/

bhalevy commented:
https://github.com/scylladb/scylla-dtest/pull/3034 is required for CI to pass for this PR

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4312/

asias commented on service/storage_service.cc:
```diff
@@ -1906,48 +1908,171 @@ future<> storage_service::do_stop_ms() {
+    const storage_service& ss;
+    sstring desc;
+    locator::host_id host_id;
+    inet_address endpoint;
```
Can we add comments explaining what are host_id and endpoint?

bhalevy replied:
In v3

asias commented on service/storage_service.cc:
```diff
@@ -1906,48 +1908,171 @@ future<> storage_service::do_stop_ms() {
+        co_await coroutine::parallel_for_each(sync_nodes, [&] (const gms::inet_address& node) -> future<> {
+            if (nodes_unknown_verb.contains(node) || nodes_down.contains(node) ||
+                    (nodes_failed.contains(node) && op_desc != "abort")) {
+                // Note that we still send to failed nodes that 
```
The comments here are not complete.

asias replied:
It looks like it is not updated in latest version?

bhalevy replied:
Fixed in v4

asias commented on service/storage_service.cc:
```diff
@@ -1906,48 +1908,171 @@ future<> storage_service::do_stop_ms() {
+            }
+        }
+        if (!nodes_down.empty()) {
+            auto msg = format("{}[{}]: nodes={} needed for {} operation are down. It is highly recommended to fix the down nodes and try again.", desc, uuid(), nodes_down, desc);
```
Nit: Most of the existing logs are in the form of
```
{}[{}]: Something
```
(First letter in capital)

instead of 

```
{}[{}]: something
```
It would be nice to be this way for new messages as well.

bhalevy replied:
No problem

bhalevy replied:
In v3

asias commented on service/storage_service.cc:
```diff
@@ -1906,48 +1908,171 @@ future<> storage_service::do_stop_ms() {
+        sstring sep = "";
+        if (!nodes_failed.empty()) {
+            msg = format("nodes={} failed {} operation", nodes_failed, op_desc);
+            sep = ";";
```
Is it better to put the error msgs in a container and join with ";"?  So no more tricks with the sep all the places.

bhalevy replied:
okay, can do.

bhalevy replied:
In v3

asias commented:
This is a nice cleanup and consolidation. IMO, it is best to merge this after we have sorted out https://github.com/scylladb/scylladb/issues/12764.

bhalevy commented:
In v3 ([7d20bac](https://github.com/scylladb/scylladb/commit/7d20bacf89335685f911762fe79ab98a6f624e1f)):
- Added comments to node_ops_ctl host_id/endpoint members
- Cleaned up error messages (while still not breaking dtests that depend on them, hopefully)
- Used `boost::algorithm::join` to join multiple error messages into one semicolon-separated string.

denesb commented on service/storage_service.cc:
```diff
@@ -1906,48 +1909,174 @@ future<> storage_service::do_stop_ms() {
+    std::unordered_set<gms::inet_address> ignore_nodes;
+    node_ops_cmd_request req;
+    abort_source as;
+    std::optional<future<>> heartbeat_updater_done_fut;
```
why are all members public?

denesb replied:
Ah, I see the control loop is still external.

bhalevy replied:
I think this future can be made private, but it's really minor.  The whole class is "private" to the current node ops functions...

denesb replied:
Yeah, I think it is fine as-is.

denesb reviewed (approved):
I'm not familiar with this code, but the restructuring looks good to me.

bhalevy commented:
@asias can you please review the latest version?

asias commented on service/storage_service.cc:
```diff
@@ -1906,48 +1909,174 @@ future<> storage_service::do_stop_ms() {
+            errors.emplace_back(format("The {} command failed on nodes={}: the needed nodes are down. It is highly recommended to fix the down nodes and try again", op_desc, nodes_failed));
         }
+        if (!errors.empty()) {
+            co_await coroutine::return_exception(std::runtime_error(boost::algorithm::join(errors, "; ")));
```
nit: can we use to_string.hh::join to avoid boost here?

bhalevy replied:
Done in v4

asias commented:
@bhalevy  I like this series.  Per https://github.com/scylladb/scylladb/pull/12799#issuecomment-1427443677, can we postpone the merge a bit? I want to avoid more changing part to the enabling RBNO by default effort.

bhalevy commented:
In v4 ([75542ec](https://github.com/scylladb/scylladb/commit/75542ec2f37ab4da74c2feb39e7d33778ad90722)):
- rebased + retested
- storage_service: add node_ops_ctl class to formalize all node_ops flow
  - Fixed comment about sending abort commands to failed nodes
  - Added `categorize_node_ops_cmd` to detect prepare/abort verbs.
  - used to_string version of `join` rather than boost's

bhalevy commented:
> This is a nice cleanup and consolidation. IMO, it is best to merge this after we have sorted out #12764.

Thanks.  The question is whether this PR can help pinpoint the root cause of #12764.
If it could, then we should consider merging it earlier rather than later (before it rots)

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4491/

bhalevy commented:
CI hit unrelated issue: https://github.com/scylladb/scylladb/issues/12629

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4497/

asias commented:
> > This is a nice cleanup and consolidation. IMO, it is best to merge this after we have sorted out #12764.
> 
> Thanks. The question is whether this PR can help pinpoint the root cause of #12764. If it could, then we should consider merging it earlier rather than later (before it rots)

#12764 is about the rpc connection is reset for unknown reason. This PR does not help much. The node ops ctrl part of this PR is not trivial. Perhaps we can merge the first 3 patches and the last 3 patches (if it does not depends on the 4th patch).

bhalevy commented:
In v5 ([e295b78](https://github.com/scylladb/scylladb/commit/e295b78510e6185068ef5b96fd82cc770acc47b7)):
- rebased and cleaned up series
- new patches:
  - repair: do_decommission_removenode_with_repair: log ignore_nodes
  - repair: node_ops_cmd_request: add print function
- storage_service: add node_ops_ctl class to formalize all node_ops flow
  - add new code before wiring it to improve reviewability
  - copy `node_ops_ctl.ignore_nodes` to req in `send_to_all` since 
- storage_service: wire node_ops_ctl for node operations

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4746/

denesb reviewed (approved):
@asias please re-review.

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4746/

This is addressed in https://github.com/scylladb/scylla-dtest/pull/3077.
Let's merge the dtest change when this PR is approved by @asias so this PR could pass next promotion.

asias commented:
> @asias please re-review.

I will take a look.

tgrabiec commented on repair/repair.cc:
```diff
@@ -1832,6 +1832,40 @@ future<> repair_service::replace_with_repair(locator::token_metadata_ptr tmptr,
     co_return co_await do_rebuild_replace_with_repair(std::move(cloned_tmptr), std::move(op), std::move(source_dc), reason, std::move(ignore_nodes));
 }
 
+node_ops_cmd_category categorize_node_ops_cmd(node_ops_cmd cmd) noexcept {
```
This shows that category and op should have been separate fields in the command from the start.

bhalevy replied:
yeah, unfortunately.  too late for that now.

tgrabiec commented on service/storage_service.cc:
```diff
@@ -1908,6 +1908,166 @@ future<> storage_service::do_stop_ms() {
+                    slogger.warn("{}[{}]: Failed to get heartbeat response from node={}", desc, uuid(), node);
+                };
+            });
+            co_await sleep_abortable(std::chrono::seconds(10), as).handle_exception([] (std::exception_ptr) {});
```
The interval is now configurable (rebase needed?).

bhalevy replied:
will fix

bhalevy replied:
in v6

tgrabiec commented on service/storage_service.cc:
```diff
@@ -3783,7 +3783,9 @@ future<> storage_service::node_ops_abort(node_ops_id ops_uuid) {
 void storage_service::node_ops_signal_abort(std::optional<node_ops_id> ops_uuid) {
-    slogger.debug("node_ops_signal_abort: ops_uuid={}", ops_uuid);
+    if (ops_uuid) {
+        slogger.warn("Node operation ops_uuid={} watchdog expired.  Signaling the operation to abort", ops_uuid);
```
Superfluous space after "."

bhalevy replied:
will fix

bhalevy replied:
in v6

bhalevy commented on service/storage_service.cc:
```diff
@@ -2259,14 +2262,18 @@ void storage_service::run_bootstrap_ops(std::unordered_set<token>& bootstrap_tok
                 slogger.debug("bootstrap[{}]: Got abort response from node={}", uuid, node);
-            });
+            } catch (...) {
+                slogger.warn("bootstrap[{}]: Abort request failed on node={}: {}", uuid, node, std::current_exception());
```
seems good to me

bhalevy replied:
in v6

tgrabiec replied:
It doesn't seem to have changed.

bhalevy replied:
it is in a later patch: "storage_service: wire node_ops_ctl for node operations"

bhalevy commented on service/storage_service.cc:
```diff
@@ -1908,6 +1908,166 @@ future<> storage_service::do_stop_ms() {
+                if (cmd_category == node_ops_cmd_category::prepare) {
+                    slogger.warn("{}[{}]: Node {} does not support the {} verb", desc, uuid(), node, op_desc);
+                } else {
+                    slogger.warn("{}[{}]: Node {} did not find ops_uuid={} or does not support the {} verb", desc, uuid(), node, uuid(), op_desc);
```
I'm not sure about that.
It seems to be a side-effect of `node_ops_cmd_check` throwing in `node_ops_cmd_handler`.

tgrabiec replied:
That sounds strange. We shouldn't return unknown_verb_error if the handler throws something else.

bhalevy replied:
I agree, but that's orthogonal to this change.
I'm not it should block it.

bhalevy commented:
In v6 ([0f07a24](https://github.com/scylladb/scylladb/commit/0f07a24889ef8d9f63f481f498dbb075ffb31360)):
- rebased on top of 2b44631ded8b9dd4db70a0cabdc8c87789698cc6
  - added `query_pending_op` function to node_ops_ctl for decommission
- changed error messages per https://github.com/scylladb/scylladb/pull/12799#discussion_r1129664447
- used config heartbeat update interval as per https://github.com/scylladb/scylladb/pull/12799#discussion_r1129713719
- removed extraneous space in warning per https://github.com/scylladb/scylladb/pull/12799#discussion_r1129734229

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4847/

bhalevy commented:
> This is addressed in [scylladb/scylla-dtest#3077](https://github.com/scylladb/scylla-dtest/pull/3077).
> Let's merge the dtest change when this PR is approved by @asias so this PR could pass next promotion.

This is addressed in https://github.com/scylladb/scylla-dtest/pull/3077.
Let's merge the dtest change when this PR is approved so this PR can pass next promotion.

bhalevy commented:
ping @tgrabiec @asias please ack so we can merge https://github.com/scylladb/scylla-dtest/pull/3077 and then this PR

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4928/

bhalevy commented:
@tgrabiec / @scylladb/scylla-maint please merge
